### PR TITLE
PP-938: Action functions do not handle interdependent value validity …

### PR DESCRIPTION
**Pull request by anamikau**
…set at the same time

#### Issue-ID

* **[PP-938](https://pbspro.atlassian.net/browse/PP-938)** 

#### Problem description
* Added the automated test for bug verification

#### Cause / Analysis
* Added the automated test for bug verification

#### Solution description
* Added the automated test for bug verification

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [ ] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [ ] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

**Comment by agrawalravi90** [comment]
The ticket's title says "qrun failing if job's formula value falls below job_sort_formula_threshold", which is very different from the title of this PR, could you please explain why?

**Comment by bhroam** [comment]
I believe the correct ticket is PP-938.

File: test/tests/functional/pbs_qmgr.py [hunk 1]
```diff
@@ -124,3 +125,30 @@ def test_listcmd_whitespaces(self):
...
                 self.server.manager(MGR_CMD_DELETE, VNODE, id=nodename)
+
+    def test_multi_attributes(self):
+        """
+        Test to verify that if multiple attributes are set
+        simultaneously and out of which one fail then none
+        will be set.
+        """
+
+        a = {'queue_type': 'e', 'enabled': 't', 'started': 't'}
```
Comment (bhroam) [inline test/tests/functional/pbs_qmgr.py]: Please spell out execution and true

File: test/tests/functional/pbs_qmgr.py [hunk 2]
```diff
@@ -124,3 +125,30 @@ def test_listcmd_whitespaces(self):
...
+
+        a = {'partition': 'bar'}
+        self.server.manager(MGR_CMD_SET, QUEUE, a, id='workq2')
+
+        a = {'queue': 'workq', 'partition': 'bar'}
+        try:
+            self.server.manager(MGR_CMD_SET, NODE, a,
+                                id=self.mom.shortname)
+        except PbsManagerError as e:
+            msg = "Partition (null) is not part of "
```
Comment (bhroam) [inline test/tests/functional/pbs_qmgr.py]: '(null)' is a bug.  You shouldn't depend on it for this test.  This test will fail once it is fixed.

File: test/tests/functional/pbs_qmgr.py [hunk 3]
```diff
@@ -124,3 +125,30 @@ def test_listcmd_whitespaces(self):
...
+            msg += "queue for node"
+            self.logger.info("looking for error, %s" % msg)
+            self.assertTrue(_msg in e.msg[0])
```
Comment (bhroam) [inline test/tests/functional/pbs_qmgr.py]: You set a variable named 'msg' and here you're checking \_msg.  I don't see \_msg set anywhere.  Did you mean msg without the underscore?

File: test/tests/functional/pbs_qmgr.py (hunk 2 above)
Comment (anamikau) [reply test/tests/functional/pbs_qmgr.py]: How do you want me to handle this? one option is to not look for the error message and second option is while PP-1073 is fixed update this test as well.

**Comment by anamikau** [comment]
updated the title. Thanks for pointing it @agrawalravi90

File: test/tests/functional/pbs_qmgr.py (hunk 2 above)
Comment (bhroam) [reply test/tests/functional/pbs_qmgr.py]: How about just searching for the partial string?  Drop the "Partition (null)" part.  You can put in a comment in the test that you are searching for a partial string due to PP-1073.  Add a comment in PP-1073 that this test needs to be updated when it is fixed.

**Comment by anamikau** [comment]
@bhroam addressed your review comments.

**Comment by bhroam** [comment]
Looks good to me.  I sign off.

**Comment by anamikau** [comment]
Thanks @bhroam . can you approve as well?

**Comment by bhroam** [comment]
I'm waiting until you have your second reviewer.  I will need to review any changes that are required.  I'll approve when it is ready for the merge.

**Comment by agrawalravi90** [comment]
I'll be your 2nd reviewer, watch out for some comments soon, thanks

File: test/tests/functional/pbs_qmgr.py [hunk 4]
```diff
@@ -43,6 +43,7 @@
 
 
 class TestQmgr(TestFunctional):
+
```
Comment (agrawalravi90) [inline test/tests/functional/pbs_qmgr.py]: why the extra line here? Is that a convention?

File: test/tests/functional/pbs_qmgr.py [hunk 5]
```diff
@@ -124,3 +125,33 @@ def test_listcmd_whitespaces(self):
             if vndef_file is not None:
                 self.mom.delete_vnodes()
                 self.server.manager(MGR_CMD_DELETE, VNODE, id=nodename)
+
+    def test_multi_attributes(self):
```
Comment (agrawalravi90) [inline test/tests/functional/pbs_qmgr.py]: So, going by the JIRA ticket, this test verifies that scenario 2 (cyclic dependency) gets handled correctly. But, don't we also need a test to verify that scenario 1 (atomicity) gets handled correctly? or do we already have such a test somewhere else?

File: test/tests/functional/pbs_qmgr.py (hunk 4 above)
Comment (anamikau) [reply test/tests/functional/pbs_qmgr.py]: I think it was added by autopep8

File: test/tests/functional/pbs_qmgr.py (hunk 5 above)
Comment (anamikau) [reply test/tests/functional/pbs_qmgr.py]: @agrawalravi90 this test covers scenario 1 as well 
line 147 a = {'queue': 'workq', 'partition': 'bar'}

above we are setting multiple attributes. Also verifying that if partition is not set then workq is also unset.

**Comment by agrawalravi90** [comment]
Thanks for clarifying Anamika, I approved the changes.

**Comment by anamikau** [comment]
@bhroam oops!! I have squashes and rebased :(

**Comment by bhroam** [comment]
@anamikau It's alright, I had been following along.  I've seen all the changes.

**Comment by bhroam** [comment]
@anamikau it looks like something snuck in.  Can you rebase again?

**Comment by anamikau** [comment]
@bhroam @agrawalravi90 all checks passed. Could one of you merge this?

**Comment by bhroam** [comment]
@anamikau, your branch is out of date with master.  You need to rebase again.  Maybe you didn't bring your master branch up to date prior to rebasing?

**Comment by anamikau** [comment]
@bhroam redid it

**Comment by anamikau** [comment]
@bhroam all checks passed. can you merge before it gets out of synch again?

**Comment by agrawalravi90** [comment]
I'm merging it in since Bhroam left for the day

**Pull request merged**